### PR TITLE
[Date Range Input] - Part 9: Obey and control the focused field

### PR DESCRIPTION
**Pull request by cmslewis**
#### Checklist

- [x] Include tests (sooooo many tests)
- [ ] ~Update documentation~ (last thing I need to do before merging, coming in a separate PR)

#### Changes proposed in this pull request:

I tried to find some lower bar of completeness so that we can release this thing ASAP, but I really couldn't find a tradeoff that made sense. **A usable `DateRangeInput` that supports both typing and click-selection really needs to obey—and control—the focused input field.** This PR makes that happen to (I think) glorious effect, and I've tested the heck out of it to make sure every little edge case works as expected.

#### Reviewers should focus on:

- [ ] The best way to explain the functionality is to let you play with it. **Test-drive the component at the preview link below, and let me know how it feels.** Pay close attention to how hovered ranges and selected ranges interact, and also pay attention to which field is in focus throughout.
- [ ] There are tests for literally every little case I could think of regarding hover ranges with respect to selected ranges. The only domain I didn't want to think about yet is how to handle invalid/out-of-range/overlapping inputs that might currently be in the field. As of this PR, the behavior there is not rigorously defined; I think clicking will just wipe invalid inputs and start from scratch. I'm personally okay with that for now. We can revisit later if we want.
- [ ] There's a _lot_ of test code. It's simple and easy to read through, but it's also fairly repetitive. I'm confident there's a way to build a parameterized suite of tests that would _dramatically_ reduce the amount of test code without reducing the footprint of the tests. LMK if you want me to spend time figuring that out before merging. I'd prefer not to for this PR in the interest of time.
- [ ] I haven't yet thought about controlled mode or adding an `onHoverChange` prop for hovering. Didn't feel necessary for the initial release of the `DateRangeInput` component. Can revisit later.
- [ ] I added a good amount of new code in `dateRangePicker.tsx`, but the merge with Ben's code in `master` (from #575) should still be easy.

#### Screenshot

(See preview for now, will add a GIF tomorrow)

### CODE REVIEW

File: packages/datetime/src/dateRangeInput.tsx [hunk 1]
```diff
@@ -180,6 +186,14 @@ export class DateRangeInput extends AbstractComponent<IDateRangeInputProps, IDat
         };
     }
 
+    public componentDidUpdate() {
+        if (this.state.isStartInputFocused && document.activeElement !== this.startInputRef) {
+            this.startInputRef.focus();
+        } else if (this.state.isEndInputFocused && document.activeElement !== this.endInputRef) {
+            this.endInputRef.focus();
```
Comment (giladgray) [inline packages/datetime/src/dateRangeInput.tsx]: should also check that `this.endInputRef !== undefined`?

File: packages/datetime/src/dateRangeInput.tsx [hunk 2]
```diff
@@ -256,14 +276,219 @@ export class DateRangeInput extends AbstractComponent<IDateRangeInputProps, IDat
...
+            } else if (isHoveredEndDefined) {
+                // we'd be replacing the end date on click
+                isStartInputFocused = false;
+                isEndInputFocused = true;
+            } else if (isHoveredStartDefined) {
+                // we'd be converting the selected end date to a start date
+                isStartInputFocused = true;
+                isEndInputFocused = false;
+            }
+        }
```
Comment (giladgray) [inline packages/datetime/src/dateRangeInput.tsx]: noting that this logic is frightening 😱 
"the cyclomatic complexity is off the charts"

File: packages/datetime/src/dateRangeInput.tsx [hunk 3]
```diff
@@ -111,36 +113,40 @@ export interface IDateRangeInputProps extends IDatePickerBaseProps, IProps {
 
 export interface IDateRangeInputState {
     isOpen?: boolean;
+    preferredBoundaryToModify?: DateRangeBoundary;
+    mostRecentlyFocusedField?: DateRangeBoundary;
```
Comment (giladgray) [inline packages/datetime/src/dateRangeInput.tsx]: `lastFocusedField`?

File: packages/datetime/src/dateRangeInput.tsx [hunk 4]
```diff
@@ -111,36 +113,40 @@ export interface IDateRangeInputProps extends IDatePickerBaseProps, IProps {
 
 export interface IDateRangeInputState {
     isOpen?: boolean;
+    preferredBoundaryToModify?: DateRangeBoundary;
```
Comment (giladgray) [inline packages/datetime/src/dateRangeInput.tsx]: is the `preferred` notion really valuable here or can we just call it `boundaryToModify`?

File: packages/datetime/src/dateRangeInput.tsx [hunk 5]
```diff
@@ -111,36 +113,40 @@ export interface IDateRangeInputProps extends IDatePickerBaseProps, IProps {
...
+    wasLastFocusChangeDueToHover?: boolean;
 };
 
 interface IStateKeysAndValuesObject {
     keys: {
+        hoverString: "startHoverString" | "endHoverString";
         inputString: "startInputString" | "endInputString";
         isInputFocused: "isStartInputFocused" | "isEndInputFocused";
         selectedValue: "selectedStart" | "selectedEnd";
     };
```
Comment (giladgray) [inline packages/datetime/src/dateRangeInput.tsx]: might it make sense to refactor state to store two instances of this object? `start` and `end`? then you wouldn't have to do all the cooky object creation below. could simply do `this.state[boundary].inputString`.

File: packages/datetime/src/dateRangeInput.tsx [hunk 6]
```diff
@@ -256,14 +276,219 @@ export class DateRangeInput extends AbstractComponent<IDateRangeInputProps, IDat
...
+
+        this.setState({
+            isStartInputFocused,
+            isEndInputFocused,
+            wasLastFocusChangeDueToHover: false,
+        });
+
+        // prevent the default focus-change behavior to avoid race conditions;
+        // we'll handle the focus change ourselves in componentDidUpdate.
+        e.preventDefault();
```
Comment (giladgray) [inline packages/datetime/src/dateRangeInput.tsx]: this should probably come first so it doesn't get missed in the future

File: packages/datetime/src/dateRangeInput.tsx [hunk 7]
```diff
@@ -256,14 +276,219 @@ export class DateRangeInput extends AbstractComponent<IDateRangeInputProps, IDat
...
         Utils.safeInvoke(this.props.onChange, selectedRange);
     }
 
+    private handleDateRangePickerHoverChange = (hoveredRange: DateRange) => {
+        if (hoveredRange == null) {
+            // undo whatever focus changes we made while hovering
+            // over various calendar dates
+            const isEndInputFocused = (this.state.preferredBoundaryToModify === DateRangeBoundary.END);
+            const isStartInputFocused = !isEndInputFocused;
+            const mostRecentlyFocusedField = (isEndInputFocused) ? DateRangeBoundary.END : DateRangeBoundary.START;
```
Comment (cmslewis) [inline packages/datetime/src/dateRangeInput.tsx]: I haaaaate that I had to set `mostRecentlyFocusedField` here. The behavior works fine to the human eye without this, but some tests were failing because `handleInputFocus` wasn't being invoked before the test assertion ran. Thus, the `is{Start,End}InputFocused` variables weren't being updated in time, so downstream code that referred to these variables was doing incorrect intermediate things before ultimately correcting itself. Ugh.

File: packages/datetime/src/dateRangeInput.tsx [hunk 8]
```diff
@@ -256,14 +276,219 @@ export class DateRangeInput extends AbstractComponent<IDateRangeInputProps, IDat
...
+                isEndInputFocused = false;
+            }
+        }
+
+        this.setState({
+            startHoverString,
+            endHoverString,
+            isStartInputFocused,
+            isEndInputFocused,
+            mostRecentlyFocusedField: (isStartInputFocused) ? DateRangeBoundary.START : DateRangeBoundary.END,
```
Comment (cmslewis) [inline packages/datetime/src/dateRangeInput.tsx]: Same.

File: packages/datetime/test/dateRangeInputTests.tsx [hunk 9]
```diff
@@ -487,6 +488,1304 @@ describe("<DateRangeInput>", () => {
...
+                        it("sets selection to [<hoveredDate>, null]", () => {
+                            assertInputTextsEqual(root, HOVER_TEST_DATE_CONFIG.str, "");
+                        });
+
+                        it("moves focus to end field", () => {
+                            assertEndInputFocused(root);
+                        });
+                    });
+
+                    describe("if mouse moves to no longer be over a calendar day...", () => {
```
Comment (giladgray) [inline packages/datetime/test/dateRangeInputTests.tsx]: no need for `...`s in describes. we know they're "nested." that's plain unconventional!

File: packages/datetime/src/dateRangeInput.tsx [hunk 10]
```diff
@@ -286,10 +511,19 @@ export class DateRangeInput extends AbstractComponent<IDateRangeInputProps, IDat
...
+        const preferredBoundaryToModify = (this.state.wasLastFocusChangeDueToHover)
+            ? this.state.preferredBoundaryToModify
+            : boundary;
+
         this.setState({
             isOpen: true,
+            preferredBoundaryToModify,
             [keys.inputString]: inputString,
             [keys.isInputFocused]: true,
+            mostRecentlyFocusedField: boundary,
```
Comment (cmslewis) [inline packages/datetime/src/dateRangeInput.tsx]: Gah, would really prefer to set this variable only in `handleInputFocus`. Don't remember off the top of my head why setting it here was necessary. Can take a look after this PR if that's okay.

File: packages/datetime/test/dateRangeInputTests.tsx [hunk 11]
```diff
@@ -487,6 +488,1304 @@ describe("<DateRangeInput>", () => {
...
+                            assertInputTextsEqual(root, "", HOVER_TEST_DATE_CONFIG.str);
+                        });
+
+                        it("moves focus to start field", () => {
+                            assertStartInputFocused(root);
+                        });
+                    });
+
+                    describe("if mouse moves to no longer be over a calendar day...", () => {
+
```
Comment (giladgray) [inline packages/datetime/test/dateRangeInputTests.tsx]: i'm personally not a fan of this leading whitespace in functions. not having it makes nested blocks a little easier to visually parse.

File: packages/datetime/src/dateRangeInput.tsx [hunk 12]
```diff
@@ -428,9 +662,12 @@ export class DateRangeInput extends AbstractComponent<IDateRangeInputProps, IDat
 
     private getInputDisplayString = (boundary: DateRangeBoundary) => {
         const { values } = this.getStateKeysAndValuesForBoundary(boundary);
-        const { isInputFocused, inputString, selectedValue } = values;
+        const { isInputFocused, inputString, selectedValue, hoverString } = values;
 
-        if (isInputFocused) {
+        if (hoverString != null && !this.isControlled()) {
```
Comment (cmslewis) [inline packages/datetime/src/dateRangeInput.tsx]: Wasn't sure exactly how to handle hover values in controlled mode. You can still see the visual hovered range in the `DateRangePicker`; maybe it makes sense to show the hovered-range values in the fields as well.

File: packages/datetime/src/dateRangeInput.tsx [hunk 13]
```diff
@@ -512,6 +753,10 @@ export class DateRangeInput extends AbstractComponent<IDateRangeInputProps, IDat
         return (boundary === DateRangeBoundary.START) ? DateRangeBoundary.END : DateRangeBoundary.START;
     }
 
+    private areSameDay = (a: moment.Moment, b: moment.Moment) => {
```
Comment (cmslewis) [inline packages/datetime/src/dateRangeInput.tsx]: Didn't use `DateUtils.areSameDay`, because that function expects`Date` parameters, not `moment.Moment`s. Don't love that this new function lives here, but given that it's not used anywhere else, I thought it was okay for now.

File: packages/datetime/src/dateRangePicker.tsx [hunk 14]
```diff
@@ -64,6 +64,15 @@ export interface IDateRangePickerProps extends IDatePickerBaseProps, IProps {
     onHoverChange?: (hoveredDates: DateRange) => void;
 
     /**
+     * The date-range boundary that the next click should modify.
```
Comment (cmslewis) [inline packages/datetime/src/dateRangePicker.tsx]: Would love some help with the language in this whole block.

File: packages/datetime/src/dateRangePicker.tsx [hunk 15]
```diff
@@ -64,6 +64,15 @@ export interface IDateRangePickerProps extends IDatePickerBaseProps, IProps {
...
+     * This will be honored unless the next click would overlap an existing date selection for the other boundary.
+     * In that case, the next click will auto-swap the two boundary dates to keep the date range boundaries in
+     * chronological order, effectively changing the other boundary's selected date.
+     * If `null`, the picker will revert to its default selection behavior.
+     */
+    preferredBoundaryToModify?: DateRangeBoundary;
```
Comment (cmslewis) [inline packages/datetime/src/dateRangePicker.tsx]: Maybe with the name of the prop too.

File: packages/datetime/src/dateRangePicker.tsx [hunk 16]
```diff
@@ -324,29 +346,86 @@ export class DateRangePicker
...
+                    const isSingleDayRangeSelected = DateUtils.areSameDay(boundaryDate, otherBoundaryDate);
+                    const nextOtherBoundaryDate = isSingleDayRangeSelected ? null : otherBoundaryDate;
+                    nextValue = this.createRangeForBoundary(null, nextOtherBoundaryDate, boundary);
+                } else if (DateUtils.areSameDay(day, otherBoundaryDate)) {
+                    const nextOtherBoundaryDate = (allowSingleDayRange) ? otherBoundaryDate : null;
+                    nextValue = this.createRangeForBoundary(day, nextOtherBoundaryDate, boundary);
+                } else if (this.isDateOverlappingOtherBoundary(day, otherBoundaryDate, boundary)) {
+                    nextValue = this.createRangeForBoundary(day, null, boundary);
+                } else {
+                    // extend the date range with an earlier boundaryDate date
```
Comment (cmslewis) [inline packages/datetime/src/dateRangePicker.tsx]: nit: `earlier` is a relic from older code. reword this to indicate we're "stretching" the range (or something).
nit: `boundaryDate date` => `boundaryDate`

File: packages/datetime/src/dateRangePicker.tsx [hunk 17]
```diff
@@ -324,29 +346,86 @@ export class DateRangePicker
...
+                    nextValue = this.createRangeForBoundary(day, otherBoundaryDate, boundary);
+                }
+            }
+        } else {
+            if (start == null && end == null) {
```
Comment (cmslewis) [inline packages/datetime/src/dateRangePicker.tsx]: This block is all existing code that i moved into an `else`.

File: packages/datetime/src/dateRangeInput.tsx (hunk 2 above)
Comment (cmslewis) [inline packages/datetime/src/dateRangeInput.tsx]: I think there's a way to whittle this down; it was mostly-working in my prototype with much less code, but there were a few nefarious edge cases. Now that things are working and the tests are in a good place, I can play with this (preferably in a future PR though).

File: packages/datetime/src/dateRangeInput.tsx (hunk 4 above)
Comment (cmslewis) [inline packages/datetime/src/dateRangeInput.tsx]: That's what I called it in the prototype implementation, but I wanted to communicate clearly that it wouldn't _always_ be respected, hence the addition of "preferred". But if we're cool letting the prop description alone clarify that, I'm cool with calling the prop `boundaryToModify`.

File: packages/datetime/src/dateRangeInput.tsx (hunk 5 above)
Comment (cmslewis) [inline packages/datetime/src/dateRangeInput.tsx]: Can I wait until a future PR to do that?

File: packages/datetime/src/dateRangeInput.tsx (hunk 1 above)
Comment (cmslewis) [inline packages/datetime/src/dateRangeInput.tsx]: Out of curiosity, when would that happen? When the component is being mounted/dismounted?

File: packages/datetime/src/dateRangeInput.tsx (hunk 13 above)
Comment (adidahiya) [inline packages/datetime/src/dateRangeInput.tsx]: how about `areSameMoment`

**Comment by llorca** [comment]
Love it! Kinda missing the placeholder change when focusing an empty field

File: packages/datetime/src/dateRangePicker.tsx [hunk 18]
```diff
@@ -64,6 +64,15 @@ export interface IDateRangePickerProps extends IDatePickerBaseProps, IProps {
     onHoverChange?: (hoveredDates: DateRange) => void;
 
     /**
+     * The date-range boundary that the next click should modify.
+     * This will be honored unless the next click would overlap an existing date selection for the other boundary.
+     * In that case, the next click will auto-swap the two boundary dates to keep the date range boundaries in
+     * chronological order, effectively changing the other boundary's selected date.
+     * If `null`, the picker will revert to its default selection behavior.
```
Comment (cmslewis) [inline packages/datetime/src/dateRangePicker.tsx]: Done

File: packages/datetime/src/dateRangeInput.tsx (hunk 4 above)
Comment (cmslewis) [inline packages/datetime/src/dateRangeInput.tsx]: Done

File: packages/datetime/src/dateRangeInput.tsx (hunk 3 above)
Comment (cmslewis) [inline packages/datetime/src/dateRangeInput.tsx]: Done

File: packages/datetime/src/dateRangeInput.tsx (hunk 6 above)
Comment (cmslewis) [inline packages/datetime/src/dateRangeInput.tsx]: Done

File: packages/datetime/test/dateRangeInputTests.tsx (hunk 9 above)
Comment (cmslewis) [inline packages/datetime/test/dateRangeInputTests.tsx]: Done

File: packages/datetime/test/dateRangeInputTests.tsx (hunk 11 above)
Comment (cmslewis) [inline packages/datetime/test/dateRangeInputTests.tsx]: Fixed (everywhere in this file.

**Comment by cmslewis** [comment]
@llorca - yep, can add that in a future PR after the feature gets in.

File: packages/datetime/src/dateRangeInput.tsx (hunk 2 above)
Comment (leebyp) [inline packages/datetime/src/dateRangeInput.tsx]: if it's easy to draw in a ascii style comment, a decision tree to show the intended logic flow would be helpful

File: packages/datetime/src/dateRangePicker.tsx (hunk 16 above)
Comment (leebyp) [inline packages/datetime/src/dateRangePicker.tsx]: same decision tree here if we can

File: packages/datetime/src/dateRangeInput.tsx (hunk 2 above)
Comment (cmslewis) [inline packages/datetime/src/dateRangeInput.tsx]: In a future PR, I'd actually like to pull this out into a `DateRangeSelectionStrategy` construct, with an API something like:

```
interface IDateRangeSelectionState {
  focusedBoundary: DateRangeBoundary;
  selectedRange: DateRange;
}

interface IDateRangeSelectionStrategy {
  /**
   * Returns the selected range that would result from clicking on the provided calendar day.
   * selectedRange - the currently selected range
   * day - the calendar date that was just hovered/clicked
   */
  getNextState(selectedRange: DateRange, day: Date) => IDateRangeSelectionState;
}
```

This would eliminate a lot of mutually dependent code between the `DateRangePicker` and the `DateRangeInput`.

File: packages/datetime/src/dateRangeInput.tsx (hunk 13 above)
Comment (cmslewis) [inline packages/datetime/src/dateRangeInput.tsx]: Would it need to be `areSameMomentDay` though? Same "moment" implies millisecond precision to me.

File: packages/datetime/src/dateRangePicker.tsx (hunk 16 above)
Comment (cmslewis) [inline packages/datetime/src/dateRangePicker.tsx]: See above! LMK if you really want it for this PR though.

File: packages/datetime/src/dateRangeInput.tsx (hunk 1 above)
Comment (cmslewis) [inline packages/datetime/src/dateRangeInput.tsx]: Just fixed anyway.

File: packages/datetime/src/dateRangeInput.tsx (hunk 2 above)
Comment (leebyp) [inline packages/datetime/src/dateRangeInput.tsx]: ah yes, i remember now, we wanted to defer until both are in a decent place

File: packages/datetime/src/dateRangeInput.tsx (hunk 12 above)
Comment (leebyp) [inline packages/datetime/src/dateRangeInput.tsx]: I think that showing the hovered-range values in the fields is fine since it's an intermediate state - very similar to how controlled text inputs also allow you type freely inside. What you **can't** do however, is actually set those values `onBlur` `onEnter` etc. - that's the point in my mind where draw the distinction for consumer code to kick in.

File: packages/datetime/src/dateRangeInput.tsx (hunk 13 above)
Comment (leebyp) [inline packages/datetime/src/dateRangeInput.tsx]: Worth splitting up `DateUtils` into `DateUtils` and `MomentUtils`? There have been many times I've had to check whether I'm looking at dates or moments in that file and try and work my way around name clashes.

File: packages/datetime/src/dateRangeInput.tsx (hunk 5 above)
Comment (giladgray) [inline packages/datetime/src/dateRangeInput.tsx]: def a future PR

File: packages/datetime/src/dateRangePicker.tsx (hunk 16 above)
Comment (leebyp) [inline packages/datetime/src/dateRangePicker.tsx]: It's fine, just wanted to flag, it should make reading and working on the tests a lot easier too. At the moment, both places are pretty crazy and un-confirmable to others not as involved. Ideally, as in the proposal above, I just want this class that I "believe" in

File: packages/datetime/src/dateRangeInput.tsx (hunk 13 above)
Comment (adidahiya) [inline packages/datetime/src/dateRangeInput.tsx]: makes sense. would do that in a separate PR though and leave this code as is
